src/platform_impl/macos/view.rs

```diff
@@ -315,10 +318,25 @@ declare_class!(
 
             // Empty string basically means that there's no preedit, so indicate that by sending
             // `None` cursor range.
+            let mut lowerbound: usize = preedit_string.len();
+            let mut upperbound: usize = lowerbound;
+            if !preedit_string.is_empty() {
+                unsafe {
+                    let new_bounds = get_actual_range_bounds(&preedit_string, selected_range);
+                    match new_bounds {
+                        None => {},
+                        Some(fetched_value) => {
+                            lowerbound = fetched_value.0;
+                            upperbound = fetched_value.1;
+                        }
+                    }
+                }
+            }
+
             let cursor_range = if preedit_string.is_empty() {
                 None
             } else {
-                Some((preedit_string.len(), preedit_string.len()))
+                Some((lowerbound, upperbound))
             };
```

Comment:
I feel like there are ways of making this code easier to read.

1. Maybe don't eagerly convert the `preedit_string` to a Rust `String` on line `288`/`296` in original code, but instead do something like the following, to get a `Id<NSString>` instead:
    ```
    // Line 288
    string.string(),
    // Line 296
    string.retain(),
    ```

    That way you at least won't have to convert from `&str` to `NSString` inside of `get_actual_range_bounds`, which will be more efficient.

2. Move the `!preedit_string.is_empty()` down into the `else` branch here, and get rid of the intermediary `lowerbound`/`upperbound` values.

Reply:
I think it is good. I am just afraid that I am just an amateur FNG dealing with Rust, and I don't want to screw it up. I have more experience with Swift but less with ObjC (and all pointer-related things).

This PR is editable by maintainers. Could you please help me apply your suggesions with your commits?